File: modules/gita-widgets/android/src/main/java/expo/modules/gitawidgets/GitaWidgetsModule.kt
package expo.modules.gitawidgets

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class GitaWidgetsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("GitaWidgets")

    val prefsName = "gita_widget"

    AsyncFunction("setWidgetData") { chapter: Int, verse: Int, sanskrit: String, english: String ->
      this@GitaWidgetsModule.persistAndBroadcast(chapter, verse, sanskrit, english, prefsName)
    }

    AsyncFunction("updateWidget") { chapter: Int, verse: Int, sanskrit: String, english: String ->
      this@GitaWidgetsModule.persistAndBroadcast(chapter, verse, sanskrit, english, prefsName)
    }
  }

  private fun persistAndBroadcast(
    chapter: Int,
    verse: Int,
    sanskrit: String,
    english: String,
    prefsName: String
  ) {
    val context = appContext.reactContext ?: return
    val prefs = context.getSharedPreferences(prefsName, Context.MODE_PRIVATE)
    prefs.edit()
      .putInt("widget_chapter", chapter)
      .putInt("widget_verse", verse)
      .putString("widget_sanskrit", sanskrit)
      .putString("widget_english", english)
      .apply()

    val manager = AppWidgetManager.getInstance(context)
    val component = ComponentName(context, "com.alphawolf.gita.widgets.VerseWidget")
    val ids = manager.getAppWidgetIds(component)
    if (ids.isNotEmpty()) {
      val intent = Intent(context, Class.forName("com.alphawolf.gita.widgets.VerseWidget"))
      intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
      intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
      context.sendBroadcast(intent)
    }
  }
}

File: modules/gita-widgets/ios/GitaWidgetsModule.swift
import ExpoModulesCore
import Foundation
import WidgetKit

private let appGroupId = "group.com.alphawolf.gita"
private let widgetKind = "VerseWidget"

public class GitaWidgetsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("GitaWidgets")

    AsyncFunction("setWidgetData") { (chapter: Int, verse: Int, sanskrit: String, english: String) in
      try GitaWidgetsModule.writeAndReload(chapter: chapter, verse: verse, sanskrit: sanskrit, english: english)
    }

    AsyncFunction("updateWidget") { (chapter: Int, verse: Int, sanskrit: String, english: String) in
      try GitaWidgetsModule.writeAndReload(chapter: chapter, verse: verse, sanskrit: sanskrit, english: english)
    }
  }

  private static func writeAndReload(chapter: Int, verse: Int, sanskrit: String, english: String) throws {
    guard let defaults = UserDefaults(suiteName: appGroupId) else {
      throw NSError(domain: "GitaWidgets", code: 1, userInfo: [NSLocalizedDescriptionKey: "App Group UserDefaults unavailable"])
    }
    defaults.set(chapter, forKey: "widget_chapter")
    defaults.set(verse, forKey: "widget_verse")
    defaults.set(sanskrit, forKey: "widget_sanskrit")
    defaults.set(english, forKey: "widget_english")
    defaults.synchronize()
    if #available(iOS 14.0, *) {
      WidgetCenter.shared.reloadTimelines(ofKind: widgetKind)
    }
  }
}
